# 第一题：使用sklearn的GaussianNB、BernoulliNB、MultinomialNB完成spambase垃圾邮件分类任务

实验内容：
1. 使用GaussianNB、BernoulliNB、MultinomialNB完成spambase邮件分类
2. 计算各自十折交叉验证的精度、查准率、查全率、F1值
3. 根据精度、查准率、查全率、F1值的实际意义以及四个值的对比阐述三个算法在spambase中的表现对比

# 1. 读取数据集

In [1]:
import numpy as np
spambase = np.loadtxt('data/spambase/spambase.data', delimiter = ",")
spamx = spambase[:, :57]
spamy = spambase[:, 57]

# 2. 导入模型

In [2]:
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# 3. 计算十折交叉验证下，GaussianNB、BernoulliNB、MultinomialNB的精度、查准率、查全率、F1值

In [3]:
# YOUR CODE HERE
# Initialize the Naive Bayes classifiers
gnb = GaussianNB()
mnb = MultinomialNB()
bnb = BernoulliNB()

# List of classifiers
classifiers = [gnb, mnb, bnb]
classifier_names = ['GaussianNB', 'MultinomialNB', 'BernoulliNB']

# Perform 10-fold cross-validation and calculate metrics for each classifier
for clf, clf_name in zip(classifiers, classifier_names):
    # Predict spam labels using cross_val_predict
    predicted_spam = cross_val_predict(clf, spamx, spamy, cv=10)

    # Calculate evaluation metrics
    accuracy = accuracy_score(spamy, predicted_spam)
    precision = precision_score(spamy, predicted_spam)
    recall = recall_score(spamy, predicted_spam)
    f1 = f1_score(spamy, predicted_spam)

    # Print the results
    print(f"Results for {clf_name}:")
    print(f"Accuracy: {accuracy:.4f}")
    print(f"Precision: {precision:.4f}")
    print(f"Recall: {recall:.4f}")
    print(f"F1 Score: {f1:.4f}")
    print("\n")

Results for GaussianNB:
Accuracy: 0.8218
Precision: 0.7004
Recall: 0.9570
F1 Score: 0.8089


Results for MultinomialNB:
Accuracy: 0.7864
Precision: 0.7324
Recall: 0.7215
F1 Score: 0.7269


Results for BernoulliNB:
Accuracy: 0.8839
Precision: 0.8813
Recall: 0.8152
F1 Score: 0.8470




###### 双击此处填写
算法|精度|查准率|查全率|F1值
-|-|-|-|-
GaussianNB|0.82|0.70|0.96|0.81
MultinomialNB|0.79|0.73|0.72|0.73
BernoulliNB|0.89|0.88|0.82|0.85

###### 双击此处填写分析

分析如下：这三个中伯努利NB的表现最好，查准率和查全率都可以，说明垃圾邮件的分类更适用于伯努利分布。